src/common/hist_util.cc

```diff
@@ -270,29 +271,34 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat, int max_num_bins) {
       #pragma omp single
       {
         p_part[0] = prev_sum;
-        for (int32_t i = 1; i < nthread; ++i) {
-          p_part[i] = p_part[i - 1] + row_ptr[rbegin + i*block_size];
+        if (batch_size < nthread) {
```

Comment:
Would this also solve your problem?
```
const size_t nthread = std::min(batch_size, omp_get_max_threads());
```

Reply:
sure - that is how i'd it initially ca904a8. then, i realized that not all thread blocks are pegged to the batch size. hence, i left the desired number of threads as-is.

Reply:
So the PR does not work if you do this, or there are performance side-effects? I am asking because I prefer simple solutions.

Reply:
the original intent was that the number of threads do not change, and i tried to preserve that intent. if the number of threads is pegged to the batch size everywhere, then this change would have preserved that intent. however, it is also pegged to the number of bins later. hence, if you'd a large number of bins and a smaller batch size, then that could be impacted with a change like this. hence, i decided to leave it as-is.

Reply:
So can't you just uniquely define the number of threads before each parallel block?

Reply:
sure, there could be multiple ways to solve this. doing this would also mean that someone would have to remember to readjust the number of threads every time a block here is added/moved; when the block logic changes to parallelize on something other than batch_size or number of bins etc.